example/z-websocket-chat/chat.eml.ml

```diff
@@ -0,0 +1,133 @@
+let home =
+  <!DOCTYPE html>
+  <html lang="en">
+  <head>
+  <title>Chat Example</title>
+  <script type="text/javascript">
+  window.onload = function () {
+      let myId = Math.random().toString();
+      let conn = new WebSocket("ws://" + window.location.host + "/websocket");
+
+      conn.onmessage = function (evt) {
+          let messages = evt.data.split('\n');
+          for (let i = 0; i < messages.length; i++) {
+              let item = document.createElement("div");
+              item.innerText = messages[i];
+              chat.appendChild(item);
+          }
+      };
+
+      let msg = document.getElementById("msg");
+      let chat = document.getElementById("chat");
+
+      document.getElementById("form").onsubmit = function () {
+          if (!conn) {
+              return false;
+          }
+          if (!msg.value) {
+              return false;
+          }
+          conn.send(myId + ":" + msg.value);
+          msg.value = "";
+          return false;
+      };
+
+  };
+  </script>
+  <style type="text/css">
+  html {
+      overflow: hidden;
+  }
+
+  body {
+      overflow: hidden;
+      padding: 0;
+      margin: 0;
+      width: 100%;
+      height: 100%;
+      background: gray;
+  }
+
+  #chat {
+      background: white;
+      margin: 0;
+      padding: 1em;
+      position: absolute;
+      top: 0.5em;
+      left: 0.5em;
+      right: 0.5em;
+      bottom: 3em;
+      overflow: auto;
+  }
+
+  #form {
+      padding: 0 0.5em 0 0.5em;
+      margin: 0;
+      position: absolute;
+      bottom: 1em;
+      left: 0px;
+      width: 100%;
+      overflow: hidden;
+  }
+
+  </style>
+  </head>
+  <body>
+  <div id="chat"></div>
+  <form id="form">
+      <input type="submit" value="Send" />
+      <input type="text" id="msg" size="64" autofocus />
+  </form>
+  </body>
+  </html>
+
+module Socket = struct
+  let connections = Hashtbl.create 5
```

Comment:
The way this hash table is used looks wrong to me. You always send to all the sockets, regardless of key, so what's the point of the key, and why accept it from user input? What if there is a collision? I can see that you might want to retain a reference to a socket cell in a collection of sockets so that you can remove it, but for that you can probably use an internally-generated incrementing id.

If these are meant to be channels, you probably need lists of sockets (or other collection) associated with each key.

Reply:
The reason for the key is to be able to clear it when the connection closes. I've started with a key from the client, because I was too fast :-), an internal reference would be better probably.

Reply:
I've changed this to an internally generated id. I've also renamed some things to make things clearer (hopefully).
The hash table is still not that useful, just easy to work with. I could also keep every connection in a (key, websocket) list and filter these, but the hash table works.